R/view_config_val_errors.R

```diff
@@ -24,120 +24,98 @@ view_config_val_errors <- function(x) {
+  errors_tbl <- remove_superfluous_enum_rows(errors_tbl)
 
+  # Get rid of unnecessarily verbose data entry when a data column is a data.frame
+  if (inherits(errors_tbl$data, "data.frame")) {
+    errors_tbl$data <- ""
+  }
 
-  # Create table ----
-  gt::gt(error_df) %>%
-    gt::tab_header(
-      title = title,
-      subtitle = subtitle
-    ) %>%
-    gt::tab_spanner(
-      label = gt::md("**Error location**"),
-      columns = error_loc_columns
-    ) %>%
-    gt::tab_spanner(
-      label = gt::md("**Schema details**"),
-      columns = c(
-        "keyword",
-        "message",
-        "schema"
-      )
-    ) %>%
-    gt::tab_spanner(
-      label = gt::md("**Config**"),
-      columns = "data"
-    ) %>%
-    gt::fmt_markdown(columns = c(
-      "instancePath",
-      "schemaPath",
-      "schema"
-    )) %>%
-    gt::tab_style(
-      style = gt::cell_text(whitespace = "pre"),
-      locations = gt::cells_body(columns = c(
-        "instancePath",
-        "schemaPath",
-        "schema"
-      ))
-    ) %>%
-    gt::tab_style(
-      style = gt::cell_text(whitespace = "pre-wrap"),
-      locations = gt::cells_body(columns = "schema")
-    ) %>%
-    gt::tab_style(
-      style = list(
-        gt::cell_fill(color = "#F9E3D6"),
-        gt::cell_text(weight = "bold")
-      ),
-      locations = gt::cells_body(
-        columns = c("message", "data")
-      )
-    ) %>%
-    gt::cols_width(
-      "schema" ~ gt::pct(1.5 / 6 * 100),
-      "data" ~ gt::pct(1 / 6 * 100),
-      "message" ~ gt::pct(1 / 6 * 100)
-    ) %>%
-    gt::cols_align(
-      align = "center",
-      columns = c(
-        "keyword",
-        "message",
-        "data"
-      )
-    ) %>%
-    gt::tab_options(
-      column_labels.font.weight = "bold",
-      table.margin.left = gt::pct(2),
-      table.margin.right = gt::pct(2),
-      data_row.padding = gt::px(5),
-      heading.background.color = "#F0F3F5",
-      column_labels.background.color = "#F0F3F5"
-    ) %>%
-    gt::tab_source_note(
-      source_note = gt::md("For more information, please consult the
-                                 [**`hubDocs` documentation**.](https://hubverse.io/en/latest/)")
-    )
-}
+  # Extract missingProperties property names from params to the data column.
+  if (any(errors_tbl$keyword == "required")) {
+    errors_tbl <- extract_params_to_data(errors_tbl, "required")
+  }
+
+  # Extract additionalProperties property names to the data column.
+  if (any(errors_tbl$keyword == "additionalProperties")) {
+    errors_tbl <- extract_params_to_data(errors_tbl, "additionalProperties")
+  }
+
+  # Remove params column
+  errors_tbl["params"] <- NULL
 
+  error_df <- split(errors_tbl, seq_len(nrow(errors_tbl))) %>%
+    purrr::map(
+      ~ unlist(.x, recursive = FALSE) %>%
+        purrr::map(~ collapse_element(.x)) %>%
+        tibble::as_tibble()
+    ) %>%
+    purrr::list_rbind() %>%
+    # split long column names
+    stats::setNames(gsub("\\.", " ", names(.)))
```

Comment:
This could benefit from explanatory comments. As I'm walking through it, I understand it this way:

```r
  # 1. split the errors_tbl row-wise
  error_df <- split(errors_tbl, seq_len(nrow(errors_tbl))) %>%
    # 2. collapse each cell into a single character
    purrr::map( # for each row, 
      ~ unlist(.x, recursive = FALSE) %>% # flatten any lists into a vector
        purrr::map(~ collapse_element(.x)) %>% # collapse each element to character
        tibble::as_tibble() # convert to tibble
    ) %>%
    # 3. concatenate the rows
    purrr::list_rbind() %>%
    # split long column names
    stats::setNames(gsub("\\.", " ", names(.)))
```

Given that this has a nested map, would it be possible to split the unlist -> tibble calls into a separate function called `flatten_error_row()` (or something similar) so that it looks something like this?

```r
  # 1. split the errors_tbl row-wise
  error_df <- split(errors_tbl, seq_len(nrow(errors_tbl))) %>%
    # 2. collapse each cell into a single character
    purrr::map(flatten_error_row) %>%
    # 3. concatenate the rows
    purrr::list_rbind() %>%
    # split long column names
    stats::setNames(gsub("\\.", " ", names(.)))
```

Reply:
Sure, I've split the code you suggested into a helper function called `flatten_error_tbl_row` in 6f019264ae3b16e868df7066402aad15e61ade31 .

Regarding the commenting, in my experience it is not necessary to comment every line of code with explanation, especially when it's just repeating in text what the code is saying itself. As far as I've been taught it is best practice to use comments sparingly and reserve them for explaining why things are happening, not what

Reply:
Thank you! I know I can be frustrating at times and slow on the re-uptake and I appreciate your patience and understanding.